lib/Dialect/FIRRTL/FIRRTLTypes.cpp

```diff
@@ -1065,6 +1115,161 @@ std::pair<uint64_t, bool> FVectorType::rootChildFieldID(uint64_t fieldID,
                         fieldID >= childRoot && fieldID <= rangeEnd);
 }
 
+//===----------------------------------------------------------------------===//
+// Enum Type
+//===----------------------------------------------------------------------===//
+
+struct circt::firrtl::detail::EnumTypeStorage : mlir::TypeStorage {
+  using KeyTy = ArrayRef<EnumType::EnumElement>;
+
+  EnumTypeStorage(KeyTy elements) : elements(elements.begin(), elements.end()) {
+    RecursiveTypeProperties props{true, false, false};
+    uint64_t fieldID = 0;
+    fieldIDs.reserve(elements.size());
+    for (auto &element : elements) {
+      auto type = element.type;
+      auto eltInfo = type.getRecursiveTypeProperties();
+      props.isPassive &= eltInfo.isPassive;
+      props.containsAnalog |= eltInfo.containsAnalog;
+      props.hasUninferredWidth |= eltInfo.hasUninferredWidth;
+      fieldID += 1;
+      fieldIDs.push_back(fieldID);
+      // Increment the field ID for the next field by the number of subfields.
+      fieldID += type.getMaxFieldID();
+    }
+    maxFieldID = fieldID;
+    recProps = props;
+    assert(!props.containsAnalog && props.isPassive);
+  }
```

Comment:
Maybe we should just disallow uninferred or non-passive (an non-analog) fields for now.  I think we can add a type verifier.

Reply:
Not for now, for ever.  I just haven't looked up how to write a type verifier :)